[WIP] feat: Add windows pipelines (#332)
kubevirt/kubevirt-tekton-tasks: templates-pipelines/windows-efi-installer/README.md

# Windows EFI Installer Pipeline

This pipeline installs Windows 11/2k22 into a new DataVolume. This DataVolume is suitable to be used as a default boot source
or golden image for Windows 11/2k22 VMs.

> This example pipeline is suitable only for windows 11/2k22 (or other windows versions which require EFI - not tested!). When using 

Reviewer: ```suggestion
This example pipeline is suitable only for Windows 11/2k22 (or other windows versions which require EFI - not tested!). When using 
```

> This example pipeline is suitable only for windows 11/2k22 (or other Windows versions which require EFI - not tested!). When using 

Reviewer: ```suggestion
This example pipeline is suitable only for Windows 11/2k22 (or other Windows versions which require EFI - not tested!). When using 
```

> this example pipeline always adjust pipeline parameters for windows version you are currently using (e.g. different template name, 
> different autoattend config map, different base image name, ...). Each windows version requires change in autounattendConfigMapName 

Reviewer: ```suggestion
this example pipeline always adjust pipeline parameters for Windows version you are currently using (e.g. different template name, 
different autoattend config map, different base image name, etc.). Each Windows version requires change in autounattendConfigMapName 
```

> parameter (e.g. using `windows2k22-autounattend` config map will not work with Windows 11 and vice versa - e.g. due to different storage 

Reviewer: ```suggestion
parameter (e.g. using `windows2k22-autounattend` ConfigMap will not work with Windows 11 and vice versa or due to different storage 
```

drivers path).

> The pipeline implements this by modifying the windows iso - extracts iso files from iso, replaces prompt bootloader with no-prompt bootloader. 

Reviewer: ```suggestion
The pipeline implements this by modifying the Windows iso - extracts iso files from iso, replaces prompt bootloader with no-prompt bootloader. 
```

This helps with automated installation of Windows in EFI boot mode. By default Windows in EFI boot mode uses a prompt bootloader, which will not 
continue with the boot process until a key is pressed. By replacing it with the non-prompt bootloader no key press is required to boot into the 

> Windows installer. Then task packs updated packages to new iso, converts it with qemu-img and replaces original iso file in PVC.

Reviewer: ```suggestion
Windows installer. Then task packs updated packages to new ISO, converts it with qemu-img and replaces original ISO file in PVC.
```

> This helps with automated installation of Windows in EFI boot mode. By default Windows in EFI boot mode uses a prompt bootloader, which will not continue with the boot process until a key is pressed. By replacing it with the non-prompt bootloader no key press is required to boot into the Windows installer. Then task packs updated packages to new ISO, converts it with qemu-img and replaces original ISO file in PVC.

Reviewer 2: ```suggestion
This helps with automated installation of Windows in EFI boot mode. By default Windows in EFI boot mode uses a prompt bootloader, which will not continue with the boot process until a key is pressed. By replacing it with the no-prompt bootloader no key press is required to boot into the Windows installer. Then task packs updated packages to new ISO, converts it with qemu-img and replaces original ISO file in PVC.
```

> After the iso is modified it creates a new VM which boots from the modified Windows installation image (ISO file). The installation of Windows is 

Reviewer: ```suggestion
After the ISO file is modified it creates a new VM which boots from the modified Windows installation image (ISO file). The installation of Windows is 
```

automatically executed and controlled by a Windows answer file. Then the pipeline will wait for the installation to complete and will delete the 
created VM while keeping the resulting DataVolume with the installed operating system. The pipeline can be customized to support different 
installation requirements.

## Prerequisites

- KubeVirt `v1.0.0`
- Tekton Pipelines `v0.44.0`
- Apply ConfigMaps with Windows sysprep (or select one which you will need) - examples can be found here: https://github.com/kubevirt/kubevirt-tekton-tasks/tree/main/pipelines/windows-efi-installer/configmaps

### Obtain Windows ISO 11 download URL

1. Go to https://www.microsoft.com/en-us/software-download/windows11.
2. Fill in the edition and `English` language (other languages need to be updated in `windows11-autounattend` ConfigMap) and go to the download page.
3. Right-click on the 64-bit download button and copy the download link. The link should be valid for 24 hours.

### Obtain Windows ISO server 2022 download URL

1. Go to https://www.microsoft.com/en-us/evalcenter/evaluate-windows-server-2022.
2. Right-click Download the ISO button.

> 3. Fill in all required informations and click on Download now button

Reviewer: ```suggestion
3. Fill in all required informations and click on Download now button.
```

> 4. Select English (United States) (other languages need to be updated in `windows2k22-autounattend` ConfigMap) - 64-bit edition ISO download

Reviewer: ```suggestion
4. Select English (United States) (other languages need to be updated in `windows2k22-autounattend` ConfigMap) - 64-bit edition ISO download.
```

   Different autounattend.xml can be also passed in a separate ConfigMap with the Pipeline parameter `autounattendConfigMapName` when creating a PipelineRun.

## Pipeline Description

```
  import-win-iso --- modify-windows-iso-file --- create-vm --- wait-for-vmi-status --- create-base-dv --- cleanup-vm
                                              |
                       create-vm-root-disk --- 
```

1. `create-vm-root-disk` creates empty DV which is used for Windows installation.
2. `import-win-iso` creates new datavolume with Windows ISO file with name defined in `isoDVName` parameter.
3. `modify-windows-iso-file` extracts imported ISO file, replaces prompt bootloader (which is used as a default one when EFI is used) 
   with no-prompt bootloader, pack the updated files back to new ISO, convert the ISO and replaces original ISO with updated one.
   Replacement of bootloader is needed to be able to automate installation of Windows versions which require EFI.
4. `create-vm` task creates a VM.
   A DV with the Windows source ISO will be attached as CD-ROM and a second empty DV will be used as installation destination.

>    A third DV with the virtio-win ISO will also be attached. (Pipeline parameter `virtioContainerDiskName`)

Reviewer: ```suggestion
   A third DV with the virtio-win ISO will also be attached (Pipeline parameter `virtioContainerDiskName`).
```

5. `wait-for-vmi-status` task waits until the VM shuts down.
6. `create-base-dv` task creates an DV with the specified name and namespace (Pipeline parameters `baseDvName` and 
   `baseDvNamespace`).
   Then it clones the second DV of the installation VM into the new DV.
7. `cleanup-vm` deletes the installer VM and all of its DVs.
8. The output artifact will be the `baseDvName`/`baseDvNamespace` DV with the basic Windows installation.
   It will boot into the Windows OOBE and needs to be setup further before it can be used.

## How to run

Pipeline runs with resolvers:
```yaml
{% for item in pipeline_runs_yaml %}
{{ item | to_nice_yaml }}---
{% endfor %}
```

## Cancelling/Deleteting pipelineRuns

> When running the example pipelines, they create temporary objects (DataVolumes, VMs, templates, ...). Each pipeline has its own clean up system which 

Reviewer: ```suggestion
When running the example pipelines, they create temporary objects (DataVolumes, VMs, templates, etc.). Each pipeline has its own clean up system which 
```

should keep the cluster clean from leftovers. In case user hard deletes or cancels running pipelineRun, the pipelineRun will not clean temporary 
objects and objects will stay in the cluster and then they have to be deleted manually. To prevent this behaviour, cancel the 
[pipelineRun gracefully](https://tekton.dev/docs/pipelines/pipelineruns/#gracefully-cancelling-a-pipelinerun). It triggers special tasks, 
which remove temporary objects and keep only result DataVolume/PVC.

> windows-efi-installer pipeline generates for each pipelineRun new source datavolume which contains imported iso. This DV has generated name and is 

Reviewer: ```suggestion
windows-efi-installer pipeline generates for each pipelineRun new source DataVolume which contains imported ISO. This DV has generated name and is 
```

> deleted after pipeline succeeds. However, the created PVC will stay in cluster, but it will have terminating state. It will wait, until pipelinRun is 

Reviewer: ```suggestion
deleted after pipeline succeeds. However, the created PVC will stay in cluster, but it will have terminating state. It will wait until pipelinRun is 
```

> deleted. This behaviour is caused by a fact, that PVC is mounted into modify-windows-iso taskRun pod and pvc can be deleted only when the pod does not 

Reviewer: ```suggestion
deleted. This behaviour is caused by the fact that, PVC is mounted into modify-windows-iso taskRun pod and PVC can be deleted only when the pod does not 
```

> WIN_URL=$(./getisourl.py | sed 's/&/\\&/g')

Reviewer 2: Same as above

6. `create-base-dv` task creates an DataVolume with the specified name and namespace (Pipeline parameters `baseDvName` and `baseDvNamespace`). Then it clones the second DataVolume of the installation VirtualMachine into the new DataVolume.
7. `cleanup-vm` deletes the installer VirtualMachine and all of its DataVolumes.
8. The output artifact will be the `baseDvName`/`baseDvNamespace` DataVolume with the basic Windows installation. It will boot into the Windows OOBE and needs to be setup further before it can be used.

## How to run

Pipeline runs with resolvers:
{% for item in pipeline_runs_yaml %}
```yaml
{% if item.metadata.generateName == "windows11-installer-run-" %}
export WIN_IMAGE_DOWNLOAD_URL=$(./getisourl.py) # see paragraph Obtaining a download URL in an automated way

{% endif %}
oc create -f - <<EOF
{{ item | to_nice_yaml }}EOF
```
{% endfor %}

### Usage in multiple namespaces

When a user defines a different namespace in e.g. `baseDvNamespace`, then the serviceAccount under which the pipeline is running will require additional permissions in that namespace. Required permissions to run task in different namespace can be found in README of each task.

> ## Cancelling/Deleteting pipelineRuns

Reviewer 2: ```suggestion
## Cancelling/Deleting PipelineRuns
```
